java/java-impl/src/com/intellij/codeInsight/FunctionalInterfaceSuggester.java

```diff
@@ -248,10 +248,10 @@ else if (expression instanceof PsiMethodReferenceExpression) {
 
                 for (int i = 0; i < targetMethodParameters.length; i++) {
                   left[i + offset] = parameters[i + offset].getType();
-                  right[i + offset] = targetMethodParameters[i].getType();
+                  right[i + offset] = result.getSubstitutor().substitute(targetMethodParameters[i].getType());
```

Comment:
looks wrong to infer substitutor to suggest type which would be used to calculate the substitutor. 

Do you want to make use of type arguments provided by the reference? This is missed indeed. It's possible to create such substitutor right here (instead of MethodReferenceResolver) and pass it as `partialSubstitutor` to another overload of `inferTypeArguments` below. 

I suggest that you write tests for the changes, just add corresponding tests to `com.intellij.java.refactoring.FunctionalInterfaceSuggesterTest`

Reply:
1. I refrain from writing tests, because I don't know what the resulting list is going to be like yet since there are lots of `@FunctionalInterface`s and there are lots of combinations
2. I removed this substitution and added `result.getSubstitutor()` as `partialSubstitutor` to `com.intellij.psi.PsiResolveHelper#inferTypeArguments(...)` and seems to be working with a known issue (I will describe it below)